File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.java
class Solution {
    private int res = 0;
    
    public int maxLength(List<String> arr) {
        List<String> candy = new ArrayList<>();
        for (String s : arr) {
            if (!hasDuplicate(s)) {
                candy.add(s);
            }
        }
        
        backtracking(candy, -1, 0);
        return res;
    }
    
    private boolean hasDuplicate(String val) {
        Set<Character> res = new HashSet<>();
        
        for (char c : val.toCharArray()) {
            if (!res.add(c)) {
                return true;
            }
        }
        return false;
    }
    
    private void backtracking(List<String> arr, int idx, int cur) {
        res = Math.max(res, countBit(cur));
        
        if (idx == arr.size()) {
            return;
        }
        
        for (int i = idx + 1; i < arr.size(); i++) {
            if ((cur&bitmask(arr.get(i))) == 0) {
                backtracking(arr, i, cur|bitmask(arr.get(i)));
            } else {
                backtracking(arr, i, cur);
            }
        }
    }
    
    private int bitmask(String str) {
        int res = 0;
        
        for (int i = 0; i < str.length(); i++) {
            res |= 1 << (str.charAt(i) - 'a');
        }
        return res;
    }
    
    private int countBit(int val) {
        res = 0;
        for (int i = 0; i < 26; i++) {
            if ((val & (1 <<i)) != 0) {
                res++;
            }
        }
        return res;
    }
}

File: 1239-maximum-length-of-a-concatenated-string-with-unique-characters/1239-maximum-length-of-a-concatenated-string-with-unique-characters.py
class Solution:    
    def maxLength(self, arr: List[str]) -> int:
        candy = []
        ans = 0
        for s in arr:
            if len(s) == len(set(s)):
                candy.append(s)

        def bi(s: str) -> int:
            cur_bit = 0
            for c in s:
                cur_bit |= 1 << (ord(c) - ord('a'))
            
            return cur_bit

        def bt(idx: int, arr: List[str], cur: int):
            nonlocal ans
            ans = max(ans, bin(cur).count('1'))

            if idx == len(arr):
                return

            
            for nxt in range(idx+1, len(arr)):
                
                new_bit = bi(arr[nxt])
                
                if cur&new_bit == 0: # 겹치는게 없음
                    bt(nxt, arr, cur|new_bit)
                else:
                    bt(nxt, arr, cur)
        
        bt(-1, candy, 0)
        
        return ans
